### ISSUE TITLE
juicebox assembly converter for juicebox 1.11

### ISSUE THREAD

**Issue by LAlvarezGlez** [phasegenomics/juicebox_scripts#24]
Hi!

I would like to know if there is a new tool to generate .fasta, .agp, and .bed files describing the modified assembly file that I obtained from JuiceBox since the tool juicebox_assembly_converter.py is not working in juicebox 1.11.

Many thanks

Lucía

**Comment by maximilianpress** [comment]
Hi Lucia, thanks for writing in.

I have checked the current version of the script and it is working for me on .assembly files output from Juicebox 1.11. Could you tell us in more detail what you tried to do and what the output looked like?

Thanks, Max

**Comment by maximilianpress** [comment]
Also: if you include your starting assembly FASTA and your modified .assembly file we can check them on our end. You are free to email them to us if you don't want to post publicly: support [at] phasegenomics.com.

**Comment by LAlvarezGlez** [comment]
Thank you so mucho for your reply!

This is is what im doing:
`python2 juicebox_assembly_converter.py -a /home/la419/94/pseudo_sha1894.1.FINAL.assembly -f /home/la419/94/pseudo_sha1894.1.fa -v`

And this is the error I am getting:

```
Processing assembly file. Details:
Assembly:                       /home/la419/94/pseudo_sha1894.1.FINAL.assembly
Fasta:                          /home/la419/94/pseudo_sha1894.1.fa
Output prefix:                  /home/la419/94/pseudo_sha1894.1.FINAL
Contig mode:                    False
Simple Chromosome Names:        False

Traceback (most recent call last):
  File "juicebox_assembly_converter.py", line 792, in <module>
    simple_chr_names=simple_chr_names)
  File "juicebox_assembly_converter.py", line 78, in process
    assembly_map, scaffolds = self._read_assembly(assembly, contig_mode=contig_mode)
  File "juicebox_assembly_converter.py", line 205, in _read_assembly
    unscaffolded_contigs.remove(assembly_map[index][0])
ValueError: list.remove(x): x not in list
```
Many thanks!

Lucía

**Comment by maximilianpress** [comment]
Hi Lucia,
Looking at the error message I would guess the script is finding some contig names in the .assembly file that are not present in the FASTA. Can you confirm that the contig names are the same between the two files? 

For example some tools such as pilon or arrow will append "|arrow" to contigs. It is important to make sure that we are using the right version of the assembly, because sometimes there are multiple versions of an assembly.

If you are able to share the files that would help to diagnose the issue more quickly.

Thanks, Max

**Comment by LAlvarezGlez** [comment]
I have just email the files! Thanks a lot!

**Comment by maximilianpress** [comment]
Hi Lucia,
I took a quick look at your files, and I think I have found the problem. The script is expecting the FASTA you provide to be the starting contigs (the input to scaffolding), not the final scaffolds (the output of scaffolding).

Specifically, I looked at your FASTA:
```
samtools faidx ~/Downloads/pseudo_sha1894.1_3.fa
head ~/Downloads/pseudo_sha1894.1_3.fa.fai 
HiC_scaffold_1	846990146	16	80	81
HiC_scaffold_2	49857993	857577555	80	81
HiC_scaffold_3	558547027	908058789	80	81
HiC_scaffold_4	231326936	1473587670	80	81
HiC_scaffold_5	604910989	1707806209	80	81
HiC_scaffold_6	425778363	2320278602	80	81
HiC_scaffold_7	182783	2751379211	80	81
HiC_scaffold_8	25223	2751564295	80	81
HiC_scaffold_9	15618	2751589850	80	81
HiC_scaffold_10	17783	2751605681	80	81
 ```
And those are final scaffolds in the FASTA I think. I then looked at your .assembly file:
```
head ~/Downloads/pseudo_sha1894.1.rawchrom.FINAL.assembly
>101:::fragment_1 1 51487
>101:::fragment_2:::debris 2 26000
>101:::fragment_3 3 91000
>101:::fragment_4:::debris 4 7000
>101:::fragment_5 5 25000
>101:::fragment_6:::debris 6 13000
>101:::fragment_7 7 103548
>101:::fragment_8:::debris 8 1000
>101:::fragment_9 9 841452
>101:::fragment_10:::debris 10 17452
```
And those appear to be a set of (broken) starting contigs, as output by Juicebox after contig breaking. The script then tries to build a mapping between the contigs in the .assembly file and the contigs in the FASTA, and fails, because they are different.

In conclusion, I think that if you pass your starting contigs as a `-f` option to the script, it should work.

Thanks, Max

**Comment by LAlvarezGlez** [comment]
Dear Max,

Thanks a lot for your amazing help. As you have told me I have tried the
same but with the starting contigs, and I am getting the same error. I may
not be using the correct assembly file? I am using the one I have created
by modifying and exporting from Juicebox and now the starting contings
fasta.

Many thanks

Lucía

**Comment by maximilianpress** [comment]
Hi Lucia, that is confusing. Can you send also the starting assembly to us so that we can investigate further?

Thanks, Max

**Comment by LAlvarezGlez** [comment]
Done!

**Comment by maximilianpress** [comment]
Ok, I see a separate issue now. It looks like there is a dummy "gap contig" in your .assembly file. Specifically, this line of your .assembly file is what I'm looking at:
```
 >hic_gap_35158 35158 500
 -9802 35158 -9804 35158 -9806 35158 -13828 35158 -13536 35158 10050 35158 11051 35158 -10553 35158 -13804 35158 -13171 35158 12483 35158 -14051 35158 14907 35158 10555 35158 -
```
Note that this gap "contig" is included as every other contig in each scaffold, whereas usually it is supposed to be implicit in the .assembly format. Is this a line that you added, or is this something coming from Juicer, etc.? My version of Juicebox (v1.11.08) does not include this line in .assembly files.

Thanks, Max

**Comment by LAlvarezGlez** [comment]
Hi Max!
I haven't added any lines, this is the JuiceBox v1.11.08 output. Should i
delete this line?

Many thanks!

Lucía

**Comment by maximilianpress** [comment]
No, deleting that line will also cause problems because the gap is referenced in all of your scaffolds. Not sure what the difference is between our JB versions, but this may be something we have to push a quick fix to account for. We will consult about what to do and update later today.

Thanks, Max

**Comment by LAlvarezGlez** [comment]
Okay, thanks a lot!

Analyzing the script it seems that the main problem is that it does not
know what do with the scaffolds that are only debris (such
as 171:::debris).

Looking forward your response

Lucía

**Comment by maximilianpress** [comment]
I think that's separate- it is definitely not handling the gaps properly because we are assuming that the .assembly file contains only contigs and not gaps. 

For example, I wrote a quick script to cleanse the .assembly of the new gapping info (caveat emptor), and got past the initial error that you came to us for. There does still appear to be a different issue with a debris contig. 

We will look in a little more detail to make sure that we understand the problem.

Here is the degapping script, which I have not tested and may have bugs:

```
#/usr/bin/env python
# write a degapped assembly to STDOUT
from __future__ import print_function
import sys

with open(sys.argv[1]) as file:
    gap_index = "not an index"
    for line in file:
        if line.startswith(">"):
            if line.startswith(">hic_gap"):
                gap_index = line.split()[1]
                print("omitting entries with index {} as gaps".format(gap_index), file=sys.stderr)
                continue
            print(line.strip())
        else:
            fields = line.split()
            degapped = [field for field in fields if field != gap_index]
            print(" ".join(degapped))

```

Here is log from using a degapped assembly:

```
python juicebox_scripts/juicebox_assembly_converter.py -a degapped.assembly  -f ~/Downloads/pseudo_sha1894.1.fa -p test   
Processing assembly file. Details:
Assembly:			degapped.assembly
Fasta:				/Users/maximilianpress/Downloads/pseudo_sha1894.1.fa
Output prefix:			test
Contig mode:			False
Simple Chromosome Names:	False

Reading sequences from /Users/maximilianpress/Downloads/pseudo_sha1894.1.fa...
....................	....................
.
Sequences read

Reading .assembly file degapped.assembly...
.assembly read

Checking for breaks listed in .assembly and making them...
Traceback (most recent call last):
  File "juicebox_scripts/juicebox_assembly_converter.py", line 834, in <module>
    simple_chr_names=simple_chr_names)
  File "juicebox_scripts/juicebox_assembly_converter.py", line 91, in process
    sequences = self._add_breaks(sequences, assembly_map)
  File "juicebox_scripts/juicebox_assembly_converter.py", line 276, in _add_breaks
    new_sequences[fragment_name] = sequences[fragment_name]
KeyError: '171:::debris'

```

**Comment by maximilianpress** [comment]
I have pushed some changes to a branch that theoretically handle everything. The issue was that a whole unfragmented contig was debris, which I think is a new Juicebox feature which people hadn't used before.

can you confirm for me that you used this feature to place whole contigs into debris?

Thanks, Max

**Comment by maximilianpress** [comment]
The branch that resolves the debris issue is here: https://github.com/phasegenomics/juicebox_scripts/tree/whole_contig_debris


You will first need to remove gaps from the assembly with the `juicebox_scripts/degap_assembly.py` script (we will add that functionality to the main workflow in time if that gapping issue comes up again). Let me know if that works for you and we can know whether to merge the branch.

Thanks for bringing this to our attention!

**Comment by LAlvarezGlez** [comment]
Hi Max!

Thank you so much, this completely solves my problem! Everything works
perfect now. I did not use any special feature to place whole contings into
debris, in fact I run everything with default parameters, so I dont know,
might it be a juicebox recent update?

Again, thanks a lot for your help!!


Lucía

**Comment by maximilianpress** [comment]
Great, glad that it worked. Thanks for bringing it up and working with us!

Cheers, Max

**Comment by maximilianpress** [comment]
@luciaalvarez95 one last quick note- the BED and AGP files are set up for the FASTA generated by our script. This may be relevant if the gaps are of different sizes in our scaffolds vs. yours. If the gaps are different sizes between the two sets of scaffolds then annotations will be frameshifted on your scaffold FASTA. Be sure to check this. You can address this by either changing the hard-coded gap size in our script or by just using the FASTA from our tool.
